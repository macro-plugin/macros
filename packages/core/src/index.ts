export { createLabeledBlock } from "./block"
export { createLabeledExpr } from "./expr"
export { createSwcPlugin, transform, transformAst, transformAsync } from "./transform"
export { print, printAsync, printExpr, printExprAsync } from "./print"
export { parse, parseAsync, parseExpr, parseType } from "./parse"
export { walk, Walker } from "./walk"
export { macro, $Macro, $LitMacro, $ExprMacro, $TmplMacro, $TypeMacro, $LabeledMacro } from "./macro"
export * from "./api"
export * from "./lib"
export * from "./types"
export * from "./utils"
