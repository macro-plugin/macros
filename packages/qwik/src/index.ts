export { qwik } from "./qwik"
export { task, vtask } from "./task"
export { resource } from "./resource"
export { store } from "./store"
export { signal } from "./signal"
export { events } from "./events"
export { computed } from "./computed"
export { link, css, scoped } from "./css"
