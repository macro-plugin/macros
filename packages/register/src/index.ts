import { transformJS, transformTS } from "./transformer"

import { addHook } from "@macro-plugin/shared"

addHook(
  transformJS,
  { extensions: [".js"] }
)

addHook(
  transformTS,
  { extensions: [".jsx", ".tsx", ".ts"] }
)
