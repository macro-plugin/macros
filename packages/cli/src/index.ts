/* eslint-disable no-eval */
/* eslint-disable @typescript-eslint/no-use-before-define */
/* eslint-disable no-console */

import { MacroOptions, parse, transform } from "@macro-plugin/core"
import type { MainOptions, ShellOptions } from "./types"
import { buildTransformOptionsSync, extractFiles, extractInput, matchPattern, picomatch, writeDts } from "@macro-plugin/shared"
import { existsSync, mkdirSync, readFileSync, rmSync, writeFile, writeFileSync } from "fs"
import path, { dirname, resolve } from "path"
import { stderr, stdin } from "process"
import { transformJS, transformTS } from "../../register/src/transformer"

import { Command } from "commander"
import { createInterface } from "readline"
import pkg from "../../core/package.json"
import { spawn } from "child_process"
import watch from "node-watch"

const CHAR_COUNT = {
  "[": 0,
  "]": 0,
  "(": 0,
  ")": 0,
  "{": 0,
  "}": 0,
}

let CURRENT_CONFIG: MacroOptions

const CACHE_FILE = path.join(process.cwd(), "<repl>.ts")

const program = new Command()

program
  .name("macros")
  .description("Macro Plugin Command Line Interface.")
  .version(pkg.version)
  .passThroughOptions()
  .argument("[files...]", "input (.js | .ts | .jsx | .tsx) files or glob patterns", [])
  .option("--init [template]", "initialize macros environment")
  .option("-c, --config <file>", "use specified config file")
  .option("-e, --eval <script>", "evaluate script")
  .option("-i, --interactive", "run an interactive Macros shell")
  .option("-m, --minify", "enable/disable minification")
  .option("-o, --output <path>", "output file name or directory")
  .option("-p, --print", "print generated code")
  .action(((files: string[], options: MainOptions & { init?: boolean | string }) => {
    if (options.init) return initialize(typeof options.init === "boolean" ? "js" : options.init)

    if (files.length > 0) return files.map(i => nodeval(i, options))

    if (options.eval) {
      try {
        print(evaluate(options.eval, options))
      } catch (e) {
        console.error(e)
      }
      return
    }

    shell(options)
  }) as () => void)

program
  .command("init")
  .description("Initialize Macros environment.")
  .option("-t, --template <template>", "initialize with template, the value could be (ts, js)", "js")
  .action((options: { template?: string }) => initialize(options.template ?? "js"))

program
  .command("emit")
  .description("Emit declaration file `macros.d.ts`.")
  .action(emitTypes)

program
  .command("dev")
  .description("Transforming codes in watch mode.")
  .argument("[files...]", "input (.js | .ts | .jsx | .tsx) files or glob patterns", [])
  .option("-r, --run", "use node run the files")
  .option("-m, --minify", "enable/disable minification")
  .option("-o, --output <path>", "output file name or directory")
  .action((files: string[], options: MainOptions) => watchFiles(files, options))

program
  .command("run")
  .description("Transform and run the entry files.")
  .argument("[files...]", "input (.js | .ts | .jsx | .tsx) files or glob patterns", [])
  .option("-m, --minify", "enable/disable minification")
  .action((files: string[], options: MainOptions) => buildFiles(files, { ...options, run: true }))

program
  .command("build")
  .description("Transforming codes in production mode.")
  .argument("[files...]", "input (.js | .ts | .jsx | .tsx) files or glob patterns", [])
  // .option("--sourcemap", "output source maps for build (default: false)")
  .option("-m, --minify", "enable/disable minification")
  .option("-r, --run", "use node run the files")
  .option("-o, --output <path>", "output file name or directory")
  .option("-w, --watch", "rebuilds when modules have changed on disk")
  .action((files: string[], options: MainOptions & { watch?: boolean }) => {
    if (options.watch) return watchFiles(files, options as MainOptions)

    buildFiles(files, options as MainOptions)
  })

program
  .command("shell")
  .description("Run an interactive Macros shell.")
  .option("-a, --ast", "interactive develop ast")
  .option("-t, --transform", "interactive show transformed code")
  .option("-m, --minify", "enable/disable minification")
  .option("-s, --swc", "enable swc(typescript|jsx) transform")
  .action(options => shell(options))

program.addHelpCommand("help [command]", "Display help for command.")

program.parse()

function isExprEnd (code: string) {
  for (let i = 0; i < code.length; i++) if (code[i] in CHAR_COUNT) CHAR_COUNT[code[i] as keyof typeof CHAR_COUNT]++

  return CHAR_COUNT["("] === CHAR_COUNT[")"] && CHAR_COUNT["["] === CHAR_COUNT["]"] && CHAR_COUNT["{"] === CHAR_COUNT["}"] && !/[.?:]\s*$/.test(code)
}

function getConfig () {
  if (!CURRENT_CONFIG) CURRENT_CONFIG = buildTransformOptionsSync({})[1]
  return CURRENT_CONFIG
}

function emitTypes () {
  const config = getConfig()
  transform("", {
    ...config,
    emitDts: true,
    onEmitDts (dts) {
      writeDts(path.resolve(config.dtsOutputPath || "./macros.d.ts"), dts)
    }
  })
}

function clearChars () {
  Object.keys(CHAR_COUNT).map(i => CHAR_COUNT[i as keyof typeof CHAR_COUNT] === 0)
}

function print (result: unknown) {
  if (result && typeof result === "object" && "type" in result) {
    console.log(JSON.stringify(result, undefined, 2))
  } else if (typeof result === "string") {
    if (result !== "") console.log(JSON.stringify(result))
  } else console.log(result)
}

/** generate macros.d.ts and create a config file if not exists */
function initialize (template: string) {
  emitTypes()

  if (matchPattern(["**/macros.config.{js,cjs,mjs,ts}"]).length === 0) { writeFileSync("./macros.config." + template, 'import { defineConfig } from "@macro-plugin/core";\n\nexport default defineConfig({\n\n})\n') }
}

function evaluate (code: string, options: ShellOptions = {}) {
  return eval(transform(code, { ...getConfig(), minify: options.minify }).code)
}

function buildFiles (input: string[], options: MainOptions) {
  extractFiles(input).map(f => nodeval(f, options))
}

function watchFiles (input: string[], options: MainOptions) {
  const { files, patts } = extractInput(input)

  const pms = patts.map(i => picomatch(i))

  console.log("Start watching...")

  watch(
    "./",
    {
      recursive: true,
      filter (f) {
        if (files.includes(resolve(f))) return true
        for (const match of pms) if (match(f)) return true
        return false
      },
    },
    (evt, name) => {
      if (evt === "update") nodeval(name, options)
    }
  )
}

function nodeval (file: string, options: MainOptions = {}) {
  const macrosOutputFile = options.output || file.replace(/\.\w+$/, ".output.js")

  if (!options.run) {
    const src = readFileSync(file).toString()

    const dir = dirname(macrosOutputFile)
    if (!existsSync(dir)) mkdirSync(dir)

    writeFile(macrosOutputFile, file.endsWith(".js") ? transformJS(src) : transformTS(src, file), (err) => {
      if (err !== null) {
        console.log(`${file} → ${macrosOutputFile}`)
      }
    })

    return
  }

  const node = spawn("node", ["-r", "@macro-plugin/core/register", file], {
    env: {
      ...process.env,
      // macrosOutputFile,
    },
  })

  node.stdout.on("data", data => {
    stdin.write(data)
  })

  node.stderr.on("data", data => {
    stderr.write(data)
  })

  // node.on("close", code => {
  // if (code === 0 && !options.print) console.log(`${file} → ${macrosOutputFile}`)
  // })
}

async function shell (options: ShellOptions, history: string[] = []) {
  const readline = createInterface({
    input: process.stdin,
    output: process.stdout,
    terminal: true,
    prompt: ">>> ",
    history,
  })

  const lines: string[] = []
  let nestedLines: string[] = []

  readline.prompt()

  readline.on("line", line => {
    if (isExprEnd(line)) {
      try {
        const current = [...nestedLines, line].join("\n")
        const code = [...lines, ...nestedLines, line].join("\n")
        if (options.ast) {
          print(parse(current, { syntax: "typescript", tsx: true }).body[0])
        } else if (options.transform) {
          console.log("`" + JSON.stringify(options.swc ? transformTS(current, "<repl>.tsx") : transformJS(current)).replace(/\\"/g, "\"").slice(1, -1) + "`")
        } else {
          const result = evaluate(code, options)
          if (!/\s*(let|const|var|function|async|class|import|export|enum|while|for|throw|interface|if|switch|private|try)\s+/.test(current)) { print(result) } // preventing the use of previous return value
        }
        lines.push(current)
      } catch (e) {
        console.error(e)
      }
      clearChars()
      nestedLines = []
      readline.setPrompt(">>> ")
    } else {
      readline.setPrompt("... ")
      nestedLines.push(line)
    }
    readline.prompt()
  })

  readline.on("close", () => {
    if (existsSync(CACHE_FILE)) rmSync(CACHE_FILE)
  })
}
