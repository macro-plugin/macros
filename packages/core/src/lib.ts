import { Expression, Identifier } from "@swc/core"
import { createExprMacro, createTmplMacro } from "./api"
import { evalAst, evalExpr, span } from "./utils"

import { printExpr } from "./print"
import { walk } from "./walk"

export var $Eval = createExprMacro("$Eval", function (args) {
  const t = args[0].type

  if (t === "StringLiteral") return this.parseExpr(evalExpr(args[0].value))
  if (t === "TemplateLiteral") {
    const track = (name: string) => this.track(name)
    walk(args[0], {
      enter (node) {
        if (node.type === "Identifier") {
          const v = track((node as Identifier).value)
          if (v && v.value) this.replace(v.value)
        }
      }
    })

    return this.parseExpr(evalExpr(evalExpr(this.printExpr(args[0]))))
  }

  if (t === "FunctionExpression" || t === "ArrowFunctionExpression") return this.parseExpr(evalExpr(`(${this.printExpr(args[0])})(${args.slice(1).map(i => this.printExpr(i)).join(",")})`))
  return this.parseExpr(evalAst(args[0]).toString())
}, "(<T>(expr: string) => T) & (<T>(expr: T) => T) & (<F extends (...args: any) => any>(expr: F, ...args: Parameters<F>) => ReturnType<F>)")

export const printAst = (ast: object) => JSON.stringify(ast).replace(/("(start|end|ctxt)":\s*)(\d+)/g, "$10")

export var $Ast = createExprMacro("$Ast", function (args) {
  return this.parseExpr(printAst(args[0].type === "StringLiteral" ? this.parseExpr(args[0].value) : args[0]))
}, '<T>(expr: T) => import("@swc/core").Expression')

export var $Env = createExprMacro("$Env", function (args, typeArgs) {
  if (args[0].type !== "StringLiteral") throw new Error("$Env macro only support StringLiteral as input.")
  const value = process.env[args[0].value] ?? ""

  if (typeArgs && typeArgs?.[0].type === "TsKeywordType") {
    switch (typeArgs[0].kind) {
      case "boolean":
        return {
          type: "BooleanLiteral",
          value: !!value,
          span
        }
      case "number":
        return {
          type: "NumericLiteral",
          value: +value,
          span
        }
    }
  }

  return {
    type: "StringLiteral",
    span,
    value
  }
}, "<R>(key: string) => R")

export var $Stringify = createExprMacro("$Stringify", function (args, typeArgs) {
  return {
    type: "StringLiteral",
    span,
    value: typeArgs && typeArgs.length > 0 ? this.printType(typeArgs[0]) : this.printExpr(args[0])
  }
}, "((expr: any) => string) & (<T>() => string)")

export const printTmpl = (strings: string[], exprs: Expression[]) => strings.reduce((query, queryPart, i) => {
  const valueExists = i < exprs.length
  const text = query + queryPart

  return valueExists ? text + printExpr(exprs[i]).code : text
}, "")

export const printRawTmpl = (strings: string[], exprs: unknown[]) => strings.reduce((query, queryPart, i) => {
  const valueExists = i < exprs.length
  const text = query + queryPart

  return valueExists ? text + exprs[i] : text
}, "")

function walkObject (obj: object | null, onEnter: (k: string, v: unknown, parent: Record<string, any>) => unknown, grand?: Record<string, any>, grandKey?: string) {
  if (obj && typeof obj === "object") {
    const allKeys = Object.keys(obj)
    for (let i = 0; i < allKeys.length; i++) {
      const k = allKeys[i]
      const v = obj[k as keyof typeof obj]

      const r = onEnter(k, v, obj)
      if (r != null && grand) {
        grand[grandKey!] = r
        break
      }

      if (typeof v === "object") {
        walkObject(v, onEnter, obj, k)
      }
    }
  }
  return obj
}

export var $Expr = createTmplMacro("$Expr", function (strings, ...exprs) {
  const exprMarkers: string[] = exprs.map((_, i) => "_macro_marker_" + i + "_")
  const options = { syntax: "typescript", tsx: true } as const
  const ast = this.parseExpr(printRawTmpl(strings, exprMarkers), options)

  walkObject(ast, (k, v, parent) => {
    if (k === "span") {
      parent[k] = span
      return
    }

    if (typeof v === "string") {
      const i = exprMarkers.findIndex((m) => m === v)
      if (i !== -1) return "__macro$$Start__" + this.printExpr(exprs[i]) + "__macro$$End__"
    }
  })

  const expr = JSON.stringify(ast, undefined, 2).replace(/("__macro\$\$Start__)|(__macro\$\$End__")/g, "").replace(/\\"/g, '"')

  return this.parseExpr(expr)
}, "import(\"@swc/core\").Expression")

export var $Quote = createTmplMacro("$Quote", function (strings, ...exprs) {
  const exprMarkers: string[] = exprs.map((_, i) => "_macro_marker_" + i + "_")
  const options = { syntax: "typescript", tsx: true } as const
  const ast = this.parse(printRawTmpl(strings, exprMarkers), options)

  walkObject(ast, (k, v, parent) => {
    if (k === "span") {
      parent[k] = span
      return
    }

    if (typeof v === "string") {
      const i = exprMarkers.findIndex((m) => m === v)
      if (i !== -1) return "__macro$$Start__" + this.printExpr(exprs[i]) + "__macro$$End__"
    }
  })

  const expr = JSON.stringify(ast, undefined, 2).replace(/("__macro\$\$Start__)|(__macro\$\$End__")/g, "").replace(/\\"/g, '"')

  return this.parseExpr(expr)
}, "(import(\"@swc/core\").Expression)[]")
