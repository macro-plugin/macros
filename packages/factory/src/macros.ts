import type { ArrayExpression, BooleanLiteral, Expression, NullLiteral } from "@swc/core"
import { createArgument, createBigIntLiteral, createBooleanLiteral, createCallExpression, createIdentifier, createInvalid, createNumericLiteral, createRegExpLiteral, createStringLiteral } from "./runtime"
import { createExprMacro, parseExpr, printExpr, span } from "@macro-plugin/core"

const spanStr = JSON.stringify(span)

export const createAst = (type: string, props: Record<string, Expression> = {}) => {
  return parseExpr(`{
    "type": "${type}",
    ${Object.entries(props).map(([k, v]) => '"' + k + '": ' + (v ? printExpr(v).code : undefined)) + ","}
    "span": ${spanStr},
  }`)
}

export const $True: BooleanLiteral = {
  type: "BooleanLiteral",
  span,
  value: true
}

export const $False: BooleanLiteral = {
  type: "BooleanLiteral",
  span,
  value: false
}

export const $Null: NullLiteral = {
  type: "NullLiteral",
  span
}

export const $Void: ArrayExpression = {
  type: "ArrayExpression",
  span,
  elements: []
}

export const $Identifier = createExprMacro("$Identifier", function ([value, optional = $False]) {
  return createAst("Identifier", { value, optional })
}, "(value: string, optional?: boolean) => import(\"@swc/core\").Identifier").proxy(createIdentifier)

export const $StringLiteral = createExprMacro("$StringLiteral", function ([value, raw]) {
  return createAst("StringLiteral", { value, raw })
}, "(value: string, raw?: string) => import(\"@swc/core\").StringLiteral").proxy(createStringLiteral)

export const $NumericLiteral = createExprMacro("$NumericLiteral", function ([value, raw]) {
  return createAst("NumericLiteral", { value, raw })
}, "(value: number, raw?: string) => import(\"@swc/core\").NumericLiteral").proxy(createNumericLiteral)

export const $BigIntLiteral = createExprMacro("$BigIntLiteral", function ([value, raw]) {
  return createAst("BigIntLiteral", { value, raw })
}, "(value: bigint, raw?: string) => import(\"@swc/core\").BigIntLiteral").proxy(createBigIntLiteral)

export const $BooleanLiteral = createExprMacro("$BooleanLiteral", function ([value]) {
  return createAst("BooleanLiteral", { value })
}, "(value: boolean) => import(\"@swc/core\").BooleanLiteral").proxy(createBooleanLiteral)

export const $RegExpLiteral = createExprMacro("$RegExpLiteral", function ([pattern, flags]) {
  return createAst("RegExpLiteral", { pattern, flags })
}, "(pattern: string, flags: string) => import(\"@swc/core\").RegExpLiteral").proxy(createRegExpLiteral)

export const $Argument = createExprMacro("$Argument", function ([expression, spread]) {
  return this.parseExpr(`{
    ${spread ? ('"spread": ' + JSON.stringify(span) + ",") : ""}
    "expression": ${this.printExpr(expression)}
  }`)
}, "(expression: import(\"@swc/core\").Expression, spread?: boolean) => import(\"@swc/core\").Argument").proxy(createArgument)

export const $CallExpression = createExprMacro("$CallExpression", function ([callee, callArgs = $Void, typeArguments]) {
  return createAst("CallExpression", {
    callee,
    arguments: callArgs,
    typeArguments
  })
}, "(callee: import(\"@swc/core\").Expression | import(\"@swc/core\").Super | import(\"@swc/core\").Import, args?: import(\"@swc/core\").Argument[], typeArguments?: import(\"@swc/core\").TsTypeParameterInstantiation) => import(\"@swc/core\").CallExpression").proxy(createCallExpression)

export const $Param = createExprMacro("$Param", function ([pat, decorators]) {
  return createAst("Parameter", { pat, decorators })
}, "(pat: import(\"@swc/core\").Pattern, decorators?: import(\"@swc/core\").Decorator[]) => import(\"@swc/core\").Param")

export const $Invalid = createExprMacro("$Invalid", function () { return createAst("Invalid") }).proxy(createInvalid)

// TODO: complete more macros for creating ast
