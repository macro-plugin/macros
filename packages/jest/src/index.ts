import { Config, createSwcPlugin, getSpanOffset } from "@macro-plugin/core"
import { Options, parse, parseSync, version as swcVersion, transform, transformSync } from "@swc/core"
import type { TransformOptions, Transformer } from "@jest/transform"

import { buildTransformOptions } from "@macro-plugin/shared"
import { createHash } from "crypto"
import getCacheKeyFunction from "@jest/create-cache-key-function"
import { readFileSync } from "fs"
import { version } from "../package.json"

function insertInstrumentOptions (jestOptions: TransformOptions<unknown>, canInstrument: boolean, inputOptions: Options, instrumentOptions?: any) {
  const shouldInstrument = jestOptions.instrument && canInstrument

  if (!shouldInstrument) {
    return inputOptions
  }

  if (inputOptions?.jsc?.experimental?.plugins?.some((x) => x[0] === "swc-plugin-coverage-instrument")) {
    return
  }

  if (!inputOptions.jsc) {
    inputOptions.jsc = {}
  }

  if (!inputOptions.jsc.experimental) {
    inputOptions.jsc.experimental = {}
  }

  if (!Array.isArray(inputOptions.jsc.experimental.plugins)) {
    inputOptions.jsc.experimental.plugins = []
  }

  inputOptions.jsc.experimental.plugins?.push(["swc-plugin-coverage-instrument", instrumentOptions ?? {}])
}

async function createTransformer (inputOptions: Config & {
  experimental?: {
    customCoverageInstrumentation?: {
      enabled: boolean
      coverageVariable?: string,
      compact?: boolean,
      reportLogic?: boolean,
      ignoreClassMethods?: Array<string>,
      instrumentLog?: { level: string, enableTrace: boolean }
    }
  }
}): Promise<Transformer> {
  const [computedSwcOptions, macroOptions, configPath] = await buildTransformOptions(inputOptions)
  const cacheKeyFunction = getCacheKeyFunction([], [swcVersion, version, JSON.stringify(inputOptions), configPath ? readFileSync(configPath).toString() : ""])
  const { enabled: canInstrument, ...instrumentOptions } = inputOptions?.experimental?.customCoverageInstrumentation ?? {}

  return {
    canInstrument: !!canInstrument, // Tell jest we'll instrument by our own
    process (src, filename, jestOptions) {
      // Determine if we actually instrument codes if jest runs with --coverage
      insertInstrumentOptions(jestOptions, !!canInstrument, computedSwcOptions, instrumentOptions)
      const offset = getSpanOffset()
      const plugin = createSwcPlugin(macroOptions, src, offset)
      const program = parseSync(src, computedSwcOptions.jsc?.parser || { syntax: "typescript", tsx: true })

      return transformSync(plugin(program), {
        ...computedSwcOptions,
        module: {
          ...computedSwcOptions.module,
          type: (jestOptions.supportsStaticESM ? "es6" : "commonjs" as any)
        },
        filename
      })
    },
    async processAsync (src, filename, jestOptions) {
      insertInstrumentOptions(jestOptions, !!canInstrument, computedSwcOptions, instrumentOptions)
      const offset = getSpanOffset()
      const plugin = createSwcPlugin(macroOptions, src, offset)
      const program = await parse(src, computedSwcOptions.jsc?.parser || { syntax: "typescript", tsx: true })

      return await transform(plugin(program), {
        ...computedSwcOptions,
        module: {
          ...computedSwcOptions.module,
          // async transform is always ESM
          type: ("es6" as any)
        },
        filename
      })
    },

    getCacheKey (src, filename, ...rest) {
      // @ts-expect-error - type overload is confused
      const baseCacheKey = cacheKeyFunction(src, filename, ...rest)

      // @ts-expect-error - signature mismatch between Jest <27 og >=27
      const options: TransformOptions = typeof rest[0] === "string" ? rest[1] : rest[0]

      return createHash("md5")
        .update(baseCacheKey)
        .update("\0", "utf8")
        .update(JSON.stringify({ supportsStaticESM: options.supportsStaticESM }))
        .digest("hex")
    }
  }
}

export default { createTransformer }
