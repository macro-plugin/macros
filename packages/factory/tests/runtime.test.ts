import { $Argument, $CallExpression, $Identifier, $StringLiteral } from "../src"

test("$Identifier macro call", () => {
  const id = $Identifier("value")

  expect(id.optional).toBe(false)
  expect(id.type).toBe("Identifier")
})

test("$StringLiteral macro call", () => {
  const str = $StringLiteral("hello")
  expect(str.value).toBe("hello")
  expect(str.raw).toBeUndefined()
})

test("$Argument macro call", () => {
  const a = $Argument($Identifier("ref"))

  expect(a).toMatchSnapshot()
})

test("$CallExpression macro call", () => {
  const a = $CallExpression($Identifier("ref"))
  const b = $CallExpression($Identifier("ref"), [
    $Argument($StringLiteral("hello")),
  ])

  expect(a).toMatchSnapshot()
  expect(b).toMatchSnapshot()
})
