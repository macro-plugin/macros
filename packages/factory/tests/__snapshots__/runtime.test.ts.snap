// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`$Argument macro call 1`] = `
{
  "expression": {
    "optional": false,
    "span": {
      "ctxt": 0,
      "end": 0,
      "start": 0,
    },
    "type": "Identifier",
    "value": "ref",
  },
  "spread": undefined,
}
`;

exports[`$CallExpression macro call 1`] = `
{
  "arguments": [],
  "callee": {
    "optional": false,
    "span": {
      "ctxt": 0,
      "end": 0,
      "start": 0,
    },
    "type": "Identifier",
    "value": "ref",
  },
  "span": {
    "ctxt": 0,
    "end": 0,
    "start": 0,
  },
  "type": "CallExpression",
  "typeArguments": undefined,
}
`;

exports[`$CallExpression macro call 2`] = `
{
  "arguments": [
    {
      "expression": {
        "raw": undefined,
        "span": {
          "ctxt": 0,
          "end": 0,
          "start": 0,
        },
        "type": "StringLiteral",
        "value": "hello",
      },
      "spread": undefined,
    },
  ],
  "callee": {
    "optional": false,
    "span": {
      "ctxt": 0,
      "end": 0,
      "start": 0,
    },
    "type": "Identifier",
    "value": "ref",
  },
  "span": {
    "ctxt": 0,
    "end": 0,
    "start": 0,
  },
  "type": "CallExpression",
  "typeArguments": undefined,
}
`;
